File: src/Forge.Host/Bridge/PayloadIO.cs
using System.Buffers.Binary;
using System.Text;

namespace Forge.Host.Bridge;

internal static class PayloadFormat
{
    public const int MaxTextBytes = 1024 * 1024;
    public const int MaxListItems = 64;

    public static void Malformed(string message) =>
        throw new BridgeProtocolException(BridgeErrorCode.MalformedPayload, message);
}

internal sealed class PayloadWriter
{
    private readonly MemoryStream _stream = new();

    public void WriteUInt32(uint value)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(bytes, value);
        _stream.Write(bytes);
    }

    public void WriteUInt64(ulong value)
    {
        Span<byte> bytes = stackalloc byte[8];
        BinaryPrimitives.WriteUInt64LittleEndian(bytes, value);
        _stream.Write(bytes);
    }

    public void WriteBoolean(bool value) => _stream.WriteByte(value ? (byte)1 : (byte)0);

    public void WriteString(string value)
    {
        ArgumentNullException.ThrowIfNull(value);
        var bytes = Encoding.UTF8.GetBytes(value);
        if (bytes.Length > PayloadFormat.MaxTextBytes) PayloadFormat.Malformed("Text field exceeds limit");
        WriteUInt32((uint)bytes.Length);
        _stream.Write(bytes);
    }

    public void WriteStrings(IReadOnlyList<string> values)
    {
        if (values.Count > PayloadFormat.MaxListItems) PayloadFormat.Malformed("List exceeds item limit");
        WriteUInt32((uint)values.Count);
        foreach (var value in values) WriteString(value);
    }

    public void WriteUInt32s(IReadOnlyList<uint> values)
    {
        if (values.Count > PayloadFormat.MaxListItems) PayloadFormat.Malformed("List exceeds item limit");
        WriteUInt32((uint)values.Count);
        foreach (var value in values) WriteUInt32(value);
    }

    public byte[] ToArray() => _stream.ToArray();
}

internal ref struct PayloadReader(ReadOnlySpan<byte> payload)
{
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);
    private readonly ReadOnlySpan<byte> _payload = payload;
    private int _offset;

    public uint ReadUInt32()
    {
        Require(4);
        var value = BinaryPrimitives.ReadUInt32LittleEndian(_payload[_offset..]);
        _offset += 4;
        return value;
    }

    public ulong ReadUInt64()
    {
        Require(8);
        var value = BinaryPrimitives.ReadUInt64LittleEndian(_payload[_offset..]);
        _offset += 8;
        return value;
    }

    public bool ReadBoolean()
    {
        Require(1);
        return _payload[_offset++] switch
        {
            0 => false,
            1 => true,
            _ => throw new BridgeProtocolException(BridgeErrorCode.MalformedPayload, "Boolean field must be zero or one"),
        };
    }

    public string ReadString()
    {
        var length = ReadUInt32();
        if (length > PayloadFormat.MaxTextBytes) PayloadFormat.Malformed("Text field exceeds limit");
        Require((int)length);
        try
        {
            var value = StrictUtf8.GetString(_payload.Slice(_offset, (int)length));
            _offset += (int)length;
            return value;
        }
        catch (DecoderFallbackException exception)
        {
            throw new BridgeProtocolException(BridgeErrorCode.MalformedPayload, $"Text field is not valid UTF-8: {exception.Message}");
        }
    }

    public string[] ReadStrings()
    {
        var count = ReadUInt32();
        if (count > PayloadFormat.MaxListItems) PayloadFormat.Malformed("List exceeds item limit");
        var values = new string[count];
        for (var index = 0; index < values.Length; index++) values[index] = ReadString();
        return values;
    }

    public uint[] ReadUInt32s()
    {
        var count = ReadUInt32();
        if (count > PayloadFormat.MaxListItems) PayloadFormat.Malformed("List exceeds item limit");
        var values = new uint[count];
        for (var index = 0; index < values.Length; index++) values[index] = ReadUInt32();
        return values;
    }

    public void Complete()
    {
        if (_offset != _payload.Length) PayloadFormat.Malformed("Payload has trailing bytes");
    }

    private void Require(int length)
    {
        if (length > _payload.Length - _offset) PayloadFormat.Malformed("Payload ended unexpectedly");
    }
}

File: src/main/bridge/PayloadIO.ts
import { Buffer } from 'node:buffer';

import { BridgeErrorCode, BridgeProtocolError } from './BridgeProtocol.ts';

const MAX_TEXT_BYTES = 1024 * 1024;
export const MAX_LIST_ITEMS = 64;

export class PayloadWriter {
  readonly #parts: Buffer[] = [];

  writeUInt32(value: number): void {
    const bytes = Buffer.allocUnsafe(4);
    bytes.writeUInt32LE(value);
    this.#parts.push(bytes);
  }

  writeUInt64(value: number): void {
    if (!Number.isSafeInteger(value) || value < 0) malformed('Invalid 64-bit integer');
    const bytes = Buffer.allocUnsafe(8);
    bytes.writeBigUInt64LE(BigInt(value));
    this.#parts.push(bytes);
  }

  writeBoolean(value: boolean): void {
    this.#parts.push(Buffer.of(value ? 1 : 0));
  }

  writeString(value: string): void {
    const bytes = Buffer.from(value, 'utf8');
    if (bytes.length > MAX_TEXT_BYTES) malformed('Text field exceeds limit');
    this.writeUInt32(bytes.length);
    this.#parts.push(bytes);
  }

  writeStrings(values: string[]): void {
    if (values.length > MAX_LIST_ITEMS) malformed('List exceeds item limit');
    this.writeUInt32(values.length);
    values.forEach((value) => this.writeString(value));
  }

  writeUInt32s(values: number[]): void {
    if (values.length > MAX_LIST_ITEMS) malformed('List exceeds item limit');
    this.writeUInt32(values.length);
    values.forEach((value) => this.writeUInt32(value));
  }

  toBuffer(): Buffer {
    return Buffer.concat(this.#parts);
  }
}

export class PayloadReader {
  readonly #bytes: Buffer;
  #offset = 0;

  constructor(payload: Uint8Array) {
    this.#bytes = Buffer.from(payload.buffer, payload.byteOffset, payload.byteLength);
  }

  readUInt32(): number {
    this.#require(4);
    const value = this.#bytes.readUInt32LE(this.#offset);
    this.#offset += 4;
    return value;
  }

  readUInt64(): number {
    this.#require(8);
    const value = Number(this.#bytes.readBigUInt64LE(this.#offset));
    this.#offset += 8;
    if (!Number.isSafeInteger(value)) malformed('64-bit integer exceeds JavaScript safe range');
    return value;
  }

  readBoolean(): boolean {
    this.#require(1);
    const value = this.#bytes[this.#offset++];
    if (value > 1) malformed('Boolean field must be zero or one');
    return value === 1;
  }

  readString(): string {
    const length = this.readUInt32();
    if (length > MAX_TEXT_BYTES) malformed('Text field exceeds limit');
    this.#require(length);
    try {
      const value = new TextDecoder('utf-8', { fatal: true })
        .decode(this.#bytes.subarray(this.#offset, this.#offset + length));
      this.#offset += length;
      return value;
    } catch {
      malformed('Text field is not valid UTF-8');
    }
  }

  readStrings(): string[] {
    const count = this.readUInt32();
    if (count > MAX_LIST_ITEMS) malformed('List exceeds item limit');
    return Array.from({ length: count }, () => this.readString());
  }

  readUInt32s(): number[] {
    const count = this.readUInt32();
    if (count > MAX_LIST_ITEMS) malformed('List exceeds item limit');
    return Array.from({ length: count }, () => this.readUInt32());
  }

  complete(): void {
    if (this.#offset !== this.#bytes.length) malformed('Payload has trailing bytes');
  }

  #require(length: number): void {
    if (length > this.#bytes.length - this.#offset) malformed('Payload ended unexpectedly');
  }
}

export function malformed(message: string): never {
  throw new BridgeProtocolError(BridgeErrorCode.MalformedPayload, message);
}
